lib/upload.js

```diff
@@ -1014,6 +1097,96 @@ function splitSizeIntoParts (totalSize, partCount) {
   return parts
 }
 
+/**
+ * Build the chunk staggers.
+ *
+ * NOTE: Must be kept in sync with `splitSizeIntoParts` in skynet-js.
+ */
+function buildChunkStaggers (totalSize, partCount, chunkSize) {
+  // The lists of chunks for each part.
+  const partsChunks = []
+  for (let i = 0; i < partCount; i++) {
+    partsChunks.push([])
+  }
+  // The leftover size that must go into the last part.
+  const hasLeftover = (totalSize % chunkSize) > 0
+
+  function newChunk (index) {
+    const partChunks = partsChunks[index]
+
+    let chunkResolve
+    const promise = new Promise((resolve) => chunkResolve = resolve)
+    const chunk = {
+      promise,
+      resolve         : chunkResolve,
+      resolved        : false,
+      nextChunkStagger: null,
+      nextChunkInPart : null,
+      indexInPart     : partChunks.length,
+    }
+
+    // Previous chunk in part should point to this chunk.
+    if (partChunks.length > 0) {
+      partChunks[partChunks.length - 1].nextChunkInPart = chunk
+    }
+
+    // Push chunk to part.
+    partChunks.push(chunk)
+  }
+
+  // Assign chunks to parts in order, looping back to the beginning if we get to
+  // the end of the parts array.
+  let lastPart = -1
+  for (let i = 0; i < Math.floor(totalSize / chunkSize); i++) {
+    newChunk(i % partCount)
+
+    if (i > lastPart) lastPart = i
+  }
+
+  if (hasLeftover) {
+    let lastIndex
+    if (lastPart === -1) {
+      // No parts were visited, so assign to the last part.
+      lastIndex = partCount - 1
+    } else {
+      // Assign the leftover to the part after the last part that was visited, or
+      // the last part in the array if all parts were used.
+      lastIndex = Math.min(lastPart + 1, partCount - 1)
+    }
+    newChunk(lastIndex)
+  }
+
+  // Now assign the stagger order. We start at the part with the most chunks, so
+  // we don't end up with any final pauses.
+  let longestPartIndex = null
+  let longestPartLength = 0
+  for (let i = 0; i < partsChunks.length; i++) {
+    if (longestPartIndex === null || partsChunks[i].length > longestPartLength) {
```

Comment:
`nit` you can get rid of the `||` if you start at `i=1` and set max to `partsChunks[0]`

Reply:
you could also have `newChunk` do it for you but that's a bit dirty, but `newChunk` is already quite dirty

Reply:
I don't understand your suggestion, can you show it in code?

> and set max to partsChunks[0]

Not sure I understand, but the last part can be the longest. Added a comment.

Reply:
it was a nit  feel free to resolve

```
let longestPartIndex = 0
let longestPartLength = partsChunks[0].lenght
for (let i = 1; ...
```
you can skip the first iteration of the loop and then also get rid of the `if (longestPartIndex === null || `

Reply:
Oh I see, thank you for the clarification!